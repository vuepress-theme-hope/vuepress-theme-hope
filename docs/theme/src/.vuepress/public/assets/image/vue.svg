<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 128 128"><path fill="#42b883" d="M78.8 10 64 35.4 49.2 10H0l64 110 64-110z"/><path fill="#35495e" d="M78.8 10 64 35.4 49.2 10H25.6L64 76l38.4-66z"/></svg>