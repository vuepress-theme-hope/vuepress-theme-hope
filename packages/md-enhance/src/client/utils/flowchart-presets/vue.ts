import base from "./base.js";

export default {
  ...base,

  // style symbol types
  symbols: {
    start: {
      class: "start-element",
      "font-color": "#fff",
      fill: "#2F495F",
      "line-width": "0px",
    },
    end: {
      class: "end-element",
      "font-color": "#fff",
      fill: "#2F495F",
      "line-width": "0px",
    },
    operation: {
      class: "operation-element",
      "font-color": "#fff",
      fill: "#00BC7D",
      "line-width": "0px",
    },
    inputoutput: {
      class: "inputoutput-element",
      "font-color": "#fff",
      fill: "#EB4D5D",
      "line-width": "0px",
    },
    subroutine: {
      class: "subroutine-element",
      "font-color": "#fff",
      fill: "#937AC4",
      "element-color": "#fff",
      "line-color": "red",
    },
    condition: {
      class: "condition-element",
      "font-color": "#fff",
      fill: "#FFB500",
      "line-width": "0px",
    },
    parallel: {
      class: "parallel-element",
      "font-color": "#fff",
      fill: "#2F495F",
      "line-width": "0px",
    },
  },
};
