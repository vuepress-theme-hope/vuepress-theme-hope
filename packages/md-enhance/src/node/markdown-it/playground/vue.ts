import { deepAssign, entries, fromEntries } from "vuepress-shared/node";

import {
  type PlaygroundData,
  type PlaygroundOptions,
  type VuePresetPlaygroundOptions,
} from "../../typings/index.js";

const VUE_SUPPORTED_EXTENSIONS = [
  "html",
  "js",
  "ts",
  "vue",
  "jsx",
  "tsx",
  "json",
];

const DEFAULT_VUE_CDN = "https://sfc.vuejs.org/vue.runtime.esm-browser.js";
const DEFAULT_VUE_SR_CDN =
  "https://sfc.vuejs.org/server-renderer.esm-browser.js";

export const getVuePlaygroundPreset = (
  options: VuePresetPlaygroundOptions = {}
): PlaygroundOptions => ({
  name: "playground#vue",
  propsGetter: (playgroundData: PlaygroundData): Record<string, string> => {
    const { title = "", files, settings: localSettings, key } = playgroundData;
    const settings = {
      // defaults
      service: "https://sfc.vuejs.org/",
      dev: false,
      ssr: false,
      ...options,
      ...localSettings,
    };

    const fileInfo = fromEntries(
      entries(files)
        .filter(([, { ext }]) => VUE_SUPPORTED_EXTENSIONS.includes(ext))
        .map(([key, { content }]) => {
          if (key === "import-map.json") {
            const importMap = <
              {
                imports: Record<string, string>;
                scopes?: Record<string, Record<string, string>>;
              }
            >JSON.parse(content);

            return [
              key,
              JSON.stringify(
                deepAssign(
                  {
                    imports: {
                      // insure vue exists
                      vue: DEFAULT_VUE_CDN,
                      // insure vue/server-renderer exists
                      ...(settings.ssr
                        ? {
                            // eslint-disable-next-line @typescript-eslint/naming-convention
                            "vue/server-renderer": DEFAULT_VUE_SR_CDN,
                          }
                        : {}),
                    },
                  },
                  importMap
                )
              ),
            ];
          }

          return [key, content];
        })
    );

    if (settings.ssr && !fileInfo["import-map.json"])
      fileInfo["import-map.json"] = JSON.stringify(
        {
          imports: {
            vue: DEFAULT_VUE_CDN,
            // eslint-disable-next-line @typescript-eslint/naming-convention
            "vue/server-renderer": DEFAULT_VUE_SR_CDN,
          },
        },
        null,
        2
      );

    return {
      key,
      title,
      link: encodeURIComponent(
        `${settings.service}#${
          // dev flag
          settings.dev ? "__DEV__" : ""
        }${
          // ssr flag
          settings.ssr ? "__SSR__" : ""
        }${
          // code base64
          Buffer.from(JSON.stringify(fileInfo)).toString("base64")
        }`
      ),
    };
  },
});
