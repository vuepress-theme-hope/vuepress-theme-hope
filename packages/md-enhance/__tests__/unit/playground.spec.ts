import MarkdownIt from "markdown-it";
import { describe, expect, it } from "vitest";

import {
  getTSPlaygroundPreset,
  getVuePlaygroundPreset,
  playground,
} from "../../src/node/markdown-it/index.js";

describe("playground", () => {
  describe("basic", () => {
    const markdownIt = MarkdownIt({ linkify: true }).use(playground, {
      name: "playground",
      component: "Playground",
      propsGetter: (data) => {
        expect(data).toMatchSnapshot();

        return {};
      },
    });

    it("Should resolve playground info", () => {
      markdownIt.render(
        `
::: playground Playground demo

@file App.vue
\`\`\`vue
<script setup>
import { ref } from 'vue'
const msg = ref('Hello World!')
</script>
<template>
  <h1>{{ msg }}</h1>
  <input v-model="msg" />
</template>
\`\`\`

@file Comp.vue

\`\`\`vue
<template>
  <div>Comp</div>
</template>
\`\`\`

@import

\`\`\`json
{
  "imports": {
    "vue": "https://sfc.vuejs.org/vue.runtime.esm-browser.js"
  }
}
\`\`\`
:::
`,
        {}
      );
    });

    it("Should resolve playground info with settings", () => {
      markdownIt.render(
        `
::: playground Playground demo2

@file App.vue

\`\`\`vue
<script setup>
import { ref } from 'vue'
const msg = ref('Hello World!')
</script>
<template>
  <h1>{{ msg }}</h1>
  <input v-model="msg" />
</template>
\`\`\`

@file Comp.vue

\`\`\`vue
<template>
  <div>Comp</div>
</template>
\`\`\`

@setting

\`\`\`json
{
  "service": "https://element-plus.run/"
}
\`\`\`

:::
`,
        {}
      );
    });

    it("Should resolve playground info with settings", () => {
      markdownIt.render(
        `
::: playground Playground demo2

@file App.vue

\`\`\`vue
<script setup>
import { ref } from 'vue'
const msg = ref('Hello World!')
</script>
<template>
  <h1>{{ msg }}</h1>
  <input v-model="msg" />
</template>
\`\`\`

@import

\`\`\`json
{
  "imports": {
    "vue": "https://sfc.vuejs.org/vue.runtime.esm-browser.js"
  }
}
\`\`\`

@setting

\`\`\`json
{
  "service": "https://element-plus.run/"
}
\`\`\`

:::
`,
        {}
      );
    });
  });

  describe("ts preset", () => {
    const markdownItWithTSPreset = MarkdownIt({ linkify: true }).use(
      playground,
      getTSPlaygroundPreset({})
    );

    it("Should work", () => {
      const result1 = markdownItWithTSPreset.render(`
::: playground#ts TS demo 1

@file index.ts

\`\`\`ts
const msg = "hello world";

const speak = (msg: string) => console.log(msg);

speak(msg);
\`\`\`

:::
`);

      const result2 = markdownItWithTSPreset.render(`
::: playground#ts TS demo 2

@file index.ts

\`\`\`ts
const msg = "hello world";

const speak = (msg: string) => console.log(msg);

speak(msg);
\`\`\`

@setting

\`\`\`json
{
  "target": "es5"
}
\`\`\`

:::
`);

      expect(result1).toMatchSnapshot();
      expect(result2).toMatchSnapshot();
    });
  });

  describe("vue preset", () => {
    const markdownItWithVuePreset = MarkdownIt({ linkify: true }).use(
      playground,
      getVuePlaygroundPreset({})
    );

    it("Should work", () => {
      const result1 = markdownItWithVuePreset.render(`
::: playground#vue Vue demo with customized imports

@file App.vue

\`\`\`vue
<script setup>
import { ref } from "vue";

import Comp from "./Comp.vue";

const msg = ref("Hello World!");
</script>

<template>
  <h1>{{ msg }}</h1>
  <input v-model="msg" />
  <Comp />
</template>
\`\`\`

@file Comp.vue

\`\`\`vue
<template>
  <div>Comp</div>
</template>
\`\`\`

@import

\`\`\`json
{
  "imports": {
    "vue": "https://sfc.vuejs.org/vue.runtime.esm-browser.js"
  }
}
\`\`\`

:::
`);
      const result2 = markdownItWithVuePreset.render(`
::: playground#vue Vue demo with customized settings

@file App.vue

\`\`\`vue
<script setup>
import { ref } from "vue";

const msg = ref("Hello Playground!");
</script>

<template>
  <h1>{{ msg }}</h1>
  <input v-model="msg" />
</template>
\`\`\`

@setting

\`\`\`json
{
  "dev": true,
  "ssr": true
}
\`\`\`

:::

`);

      expect(result1).toMatchSnapshot();
      expect(result2).toMatchSnapshot();
    });
  });
});
