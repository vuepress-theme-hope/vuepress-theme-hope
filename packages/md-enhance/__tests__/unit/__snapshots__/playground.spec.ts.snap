// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`playground > basic > Should resolve playground info 1`] = `
{
  "files": {
    "App.vue": {
      "content": "<script setup>
import { ref } from 'vue'
const msg = ref('Hello World!')
</script>
<template>
  <h1>{{ msg }}</h1>
  <input v-model=\\"msg\\" />
</template>
",
      "ext": "vue",
    },
    "Comp.vue": {
      "content": "<template>
  <div>Comp</div>
</template>
",
      "ext": "vue",
    },
    "import-map.json": {
      "content": "{
  \\"imports\\": {
    \\"vue\\": \\"https://sfc.vuejs.org/vue.runtime.esm-browser.js\\"
  }
}
",
      "ext": "json",
    },
  },
  "importMap": "import-map.json",
  "key": "62b3bf5a",
  "settings": {},
  "title": "Playground%20demo",
}
`;

exports[`playground > basic > Should resolve playground info with settings 1`] = `
{
  "files": {
    "App.vue": {
      "content": "<script setup>
import { ref } from 'vue'
const msg = ref('Hello World!')
</script>
<template>
  <h1>{{ msg }}</h1>
  <input v-model=\\"msg\\" />
</template>
",
      "ext": "vue",
    },
    "Comp.vue": {
      "content": "<template>
  <div>Comp</div>
</template>
",
      "ext": "vue",
    },
  },
  "key": "1877a7d0",
  "settings": {
    "service": "https://element-plus.run/",
  },
  "title": "Playground%20demo2",
}
`;

exports[`playground > basic > Should resolve playground info with settings 2`] = `
{
  "files": {
    "App.vue": {
      "content": "<script setup>
import { ref } from 'vue'
const msg = ref('Hello World!')
</script>
<template>
  <h1>{{ msg }}</h1>
  <input v-model=\\"msg\\" />
</template>
",
      "ext": "vue",
    },
    "import-map.json": {
      "content": "{
  \\"imports\\": {
    \\"vue\\": \\"https://sfc.vuejs.org/vue.runtime.esm-browser.js\\"
  }
}
",
      "ext": "json",
    },
  },
  "importMap": "import-map.json",
  "key": "1877a7d0",
  "settings": {
    "service": "https://element-plus.run/",
  },
  "title": "Playground%20demo2",
}
`;

exports[`playground > ts preset > Should work 1`] = `
"<Playground key=\\"2d9205c7\\" title=\\"TS%20demo%201\\" link=\\"https%3A%2F%2Fwww.typescriptlang.org%2Fplay%23code%2FMYewdgzgLgBAthA5jAvDARACwKYBtcgwDuIATrgCboDcAULaJLBAA7YCGA1qjABQKIAXDGikAlmEQBKVAD4YjCCFzYAdAUT8kUurVYdOW6XSA\\">
</Playground>
"
`;

exports[`playground > ts preset > Should work 2`] = `
"<Playground key=\\"2d9205c8\\" title=\\"TS%20demo%202\\" link=\\"https%3A%2F%2Fwww.typescriptlang.org%2Fplay%3F1%23code%2FMYewdgzgLgBAthA5jAvDARACwKYBtcgwDuIATrgCboDcAULaJLBAA7YCGA1qjABQKIAXDGikAlmEQBKVAD4YjCCFzYAdAUT8kUurVYdOW6XSA\\">
</Playground>
"
`;

exports[`playground > vue preset > Should work 1`] = `
"<Playground key=\\"68440394\\" title=\\"Vue%20demo%20with%20customized%20imports\\" link=\\"https%3A%2F%2Fsfc.vuejs.org%2F%23eyJBcHAudnVlIjoiPHNjcmlwdCBzZXR1cD5cbmltcG9ydCB7IHJlZiB9IGZyb20gXCJ2dWVcIjtcblxuaW1wb3J0IENvbXAgZnJvbSBcIi4vQ29tcC52dWVcIjtcblxuY29uc3QgbXNnID0gcmVmKFwiSGVsbG8gV29ybGQhXCIpO1xuPC9zY3JpcHQ%2BXG5cbjx0ZW1wbGF0ZT5cbiAgPGgxPnt7IG1zZyB9fTwvaDE%2BXG4gIDxpbnB1dCB2LW1vZGVsPVwibXNnXCIgLz5cbiAgPENvbXAgLz5cbjwvdGVtcGxhdGU%2BXG4iLCJDb21wLnZ1ZSI6Ijx0ZW1wbGF0ZT5cbiAgPGRpdj5Db21wPC9kaXY%2BXG48L3RlbXBsYXRlPlxuIiwiaW1wb3J0LW1hcC5qc29uIjoie1wiaW1wb3J0c1wiOntcInZ1ZVwiOlwiaHR0cHM6Ly9zZmMudnVlanMub3JnL3Z1ZS5ydW50aW1lLmVzbS1icm93c2VyLmpzXCJ9fSJ9\\">
</Playground>
"
`;

exports[`playground > vue preset > Should work 2`] = `
"<Playground key=\\"0eebfebd\\" title=\\"Vue%20demo%20with%20customized%20settings\\" link=\\"https%3A%2F%2Fsfc.vuejs.org%2F%23__DEV____SSR__eyJBcHAudnVlIjoiPHNjcmlwdCBzZXR1cD5cbmltcG9ydCB7IHJlZiB9IGZyb20gXCJ2dWVcIjtcblxuY29uc3QgbXNnID0gcmVmKFwiSGVsbG8gUGxheWdyb3VuZCFcIik7XG48L3NjcmlwdD5cblxuPHRlbXBsYXRlPlxuICA8aDE%2Be3sgbXNnIH19PC9oMT5cbiAgPGlucHV0IHYtbW9kZWw9XCJtc2dcIiAvPlxuPC90ZW1wbGF0ZT5cbiIsImltcG9ydC1tYXAuanNvbiI6IntcbiAgXCJpbXBvcnRzXCI6IHtcbiAgICBcInZ1ZVwiOiBcImh0dHBzOi8vc2ZjLnZ1ZWpzLm9yZy92dWUucnVudGltZS5lc20tYnJvd3Nlci5qc1wiLFxuICAgIFwidnVlL3NlcnZlci1yZW5kZXJlclwiOiBcImh0dHBzOi8vc2ZjLnZ1ZWpzLm9yZy9zZXJ2ZXItcmVuZGVyZXIuZXNtLWJyb3dzZXIuanNcIlxuICB9XG59In0%3D\\">
</Playground>
"
`;
